browser/ai_chat/code_execution_tool.cc

```diff
@@ -0,0 +1,207 @@
+};
+
+CodeExecutionTool::CodeExecutionRequest::CodeExecutionRequest(
+    std::unique_ptr<content::WebContents> web_contents,
+    std::string wrapped_js,
+    UseToolCallback callback)
+    : web_contents(std::move(web_contents)),
+      wrapped_js(std::move(wrapped_js)),
+      callback(std::move(callback)) {}
+
+CodeExecutionTool::CodeExecutionRequest::~CodeExecutionRequest() = default;
+
+CodeExecutionTool::CodeExecutionTool(content::BrowserContext* browser_context)
+    : profile_(Profile::FromBrowserContext(browser_context)) {}
+
+CodeExecutionTool::~CodeExecutionTool() = default;
+
+std::string_view CodeExecutionTool::Name() const {
+  return mojom::kCodeExecutionToolName;
+}
+
+std::string_view CodeExecutionTool::Description() const {
+  return "Execute JavaScript code and return a human-readable formatted "
+         "string as a result. "
+         "Do not use console.log statements or similar statements. Always "
+         "return a string as a result."
+         "The code will be executed in a sandboxed environment.";
+}
+
+std::optional<base::Value::Dict> CodeExecutionTool::InputProperties() const {
+  return CreateInputProperties(
+      {{"script", StringProperty("The JavaScript code to execute")}});
+}
+
+std::optional<std::vector<std::string>> CodeExecutionTool::RequiredProperties()
+    const {
+  return std::vector<std::string>{"script"};
+}
+
+std::variant<bool, mojom::PermissionChallengePtr>
+CodeExecutionTool::RequiresUserInteractionBeforeHandling(
+    const mojom::ToolUseEvent& tool_use) const {
+  return false;
+}
+
+bool CodeExecutionTool::SupportsConversation(
+    bool is_temporary,
+    bool has_untrusted_content,
+    mojom::ConversationCapability conversation_capability) const {
+  // Support all conversation types for now
+  return true;
+}
+
+void CodeExecutionTool::SetExecutionTimeLimitForTesting(
+    base::TimeDelta time_limit) {
+  execution_time_limit_for_testing_ = time_limit;
+}
+
+void CodeExecutionTool::UseTool(const std::string& input_json,
+                                UseToolCallback callback) {
+  auto input_dict = base::JSONReader::ReadDict(
+      input_json, base::JSON_PARSE_CHROMIUM_EXTENSIONS);
+  if (!input_dict.has_value()) {
+    std::move(callback).Run(CreateContentBlocksForText(
+        "Error: Invalid JSON input, input must be a JSON object"));
+    return;
+  }
+
+  const std::string* script = input_dict->FindString("script");
+
+  if (!script || script->empty()) {
+    std::move(callback).Run(
+        CreateContentBlocksForText("Error: Missing or empty 'script' field"));
+    return;
+  }
+
+  auto wrapped_js =
+      base::StrCat({"(async function() { try {", *script,
+                    "} catch (error) { return error.toString(); } })()"});
+
+  auto* otr_profile =
+      profile_->GetPrimaryOTRProfile(true /* create_if_needed */);
+  content::WebContents::CreateParams create_params(otr_profile);
+  auto web_contents = content::WebContents::Create(create_params);
```

Comment:
the fact that this creates WebContents for each request is killing me. Can we wire it with the conversation so it can reuse the same WebContents for the chat session?

Currently this still feels a PoC-level quality, i.e. not something that should be considered a production-ready implementation.

Saying that, I am fine if you will rework this in a next PR to make this a bit more sane and less resource-intensive.